File: POH7/Mizugi/Main.clj
; Try POH
; author: Leonardone @ NEETSDKASU
(def gs read-line)
(defn gi [] (read-string (gs)))
(defn gss [] (list* (.split (gs) " "))) ; gisのmapがlistを返すのでgssも揃える
(defn gis [] (map read-string (gss)))
(defn ngt [n f] (reverse (loop [c n r []] (if (= c 0) r (recur (dec c) (cons (f) r))))))  ; forやmapだと遅延評価で読み込み順が狂うう
(defn ngs [n] (ngt n gs))
(defn ngi [n] (ngt n gi))
(defn ngss [n] (ngt n gss))
(defn ngis [n] (ngt n gis))
; ============================================

(defn modval [v] (mod v 1000000000))
(defn reject [d v c]
    (loop [v v c c]
        (if (= (mod v d) 0) (recur (/ v d) (inc c))
            [v c]
        )))
(defn solve [n]
    (loop [v n r 1 c 0]
        (if (= v 1) (loop [r r c c] (if (= c 0) r (recur (modval (* r 2)) (dec c))))
            (let [[x c1] (reject 5 v 0)
                  [y c2] (reject 2 x (- c1))
                 ]
                (recur (dec v) (modval (* r y)) (+ c c2))
            ))))

; ============================================

(defn main []
    (let [n (gi) ans (solve n)]
        (println ans)
    ))

; ============================================
(main)
; ============================================

File: POH7/Mizugi/Main.pl
# Try POH
# author: Leonardone @ NEETSDKASU
# -----------------------------------------------------------------------------
sub gs { chomp(my $v = <>); $v }
sub gi { int gs }
sub gss { split / /, gs  }
sub gis { map int, gss }
sub ngt { my ($n, $f) = @_; my @r = (); for (1 .. $n) { push @r, &$f } @r }
sub ngs { ngt shift @_, \&gs }
sub ngi { ngt shift @_, \&gi }
sub ngss { ngt shift @_, sub { [&gss] } }
sub ngis { ngt shift @_, sub { [&gis] } }
# -----------------------------------------------------------------------------


sub solve {
    my $n = gi;
    my $r = 1;
    my $c = 0;
    for $x (1 .. $n) {
        while ($x % 5 == 0) {
            $x /= 5;
            $c--;
        }
        while ($x % 2 == 0) {
            $x /= 2;
            $c ++;
        }
        $r = ($r * $x) % 1000000000;
    }
    while ($c-- > 0) {
        $r = ($r * 2) % 1000000000
    }    
    print "$r\n";
}


solve;
